pkg/pillar/cmd/zedrouter/dnsmasq_test.go

```diff
@@ -0,0 +1,526 @@
+			dr.exitErr = e
+		}
+		dnsmasqOutputChan <- dr
+	}()
+
+	return endChan, dnsmasqOutputChan
+}
+
+func requestDhcp(mac net.HardwareAddr, vethPeerName string, timeout time.Duration) *net.IP {
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	dhcpOptions := []nclient4.ClientOpt{
+		nclient4.WithRetry(4),
+		nclient4.WithHWAddr(mac),
+		nclient4.WithTimeout(timeout),
+	}
+
+	if testing.Verbose() {
+		dhcpOptions = append(dhcpOptions, nclient4.WithSummaryLogger())
+	}
+
+	client, err := nclient4.New(vethPeerName, dhcpOptions...)
+	if err != nil {
+		panic(err)
+	}
+	defer client.Close()
+
+	lease, err := client.Request(ctx)
+	if err != nil {
+		//panic(err)
+		return nil
+	}
+
+	if lease != nil && lease.Offer != nil {
+		return &lease.Offer.YourIPAddr
+	} else {
+		return nil
+	}
+}
+
+func incByteArray(arr []byte) {
+	changed := false
+	for i := len(arr) - 1; i >= 0; i-- {
+		// byte is just an alias for uint8 (https://go.dev/ref/spec#Numeric_types)
+		if arr[i] == math.MaxUint8 {
+			arr[i] = 0
+			continue
+		}
+		arr[i]++
+		changed = true
+		break
+	}
+
+	if !changed {
+		for i := range arr {
+			arr[i] = 0
+		}
+	}
+}
+
+func multiRequestDhcp(macs []net.HardwareAddr, vethPeerName string, timeout time.Duration) mac2Ip {
+	var m2ip mac2Ip
+	var m2ipMutex sync.Mutex
+
+	var wg sync.WaitGroup
+	for _, mac := range macs {
+		wg.Add(1)
+		go func(mac net.HardwareAddr) {
+			ip := requestDhcp(mac, vethPeerName, timeout)
+			m2ipMutex.Lock()
+			if ip != nil {
+				m2ip.add(mac, *ip)
+			}
+			m2ipMutex.Unlock()
+			wg.Done()
+		}(mac)
+	}
+
+	wg.Wait()
+
+	return m2ip
+}
+
+func TestIncByteArray(t *testing.T) {
+	t.Parallel()
+
+	v := []byte{1, 2}
+	incByteArray(v)
+
+	if v[0] != 1 || v[1] != 3 {
+		t.Fatalf("incByteArray failed: %+v", v)
+	}
+
+	v = []byte{255, 255}
+	incByteArray(v)
+
+	if v[0] != 0 || v[1] != 0 {
+		t.Fatalf("incByteArray failed: %+v", v)
+	}
+
+	v = []byte{250, 255}
+	incByteArray(v)
+	if v[0] != 251 || v[1] != 0 {
+		t.Fatalf("incByteArray failed: %+v", v)
+	}
+}
+
+type dhcpNetworkEnv struct {
+	vethName          string
+	vethPeerName      string
+	bridgeName        string
+	pathToDnsmasqConf string
+}
+
+// create different interfaces if running tests in parallel
+var networkInterfacesID uint32
+
+func createDhcpNetworkEnv(bridgeAddr *netlink.Addr) (dhcpNetworkEnv, bool) {
+	var dnEnv dhcpNetworkEnv
+
+	networkInterfaceID := atomic.AddUint32(&networkInterfacesID, 1)
+
+	dnEnv.vethName = fmt.Sprintf("veth-%d", networkInterfaceID)
+	dnEnv.vethPeerName = fmt.Sprintf("vpeer-%d", networkInterfaceID)
+	dnEnv.bridgeName = fmt.Sprintf("br-%d", networkInterfaceID)
+
+	la := netlink.NewLinkAttrs()
+	la.Name = dnEnv.bridgeName
+	dnsmasqBridge := &netlink.Bridge{LinkAttrs: la}
+	err := netlink.LinkAdd(dnsmasqBridge)
+	if err != nil && err.Error() == "operation not permitted" {
+		return dnEnv, false
+	} else if err != nil {
+		panic(err)
+	}
+	err = netlink.AddrAdd(dnsmasqBridge, bridgeAddr)
```

Comment:
In general, I'm bit wary when I see unit tests interfering with the host system, e.g. by creating interfaces and assigning IP addresses that may collide with the host networking. IMHO there is a limit in how far unit tests should go and I would prefer to use them only to cover "our logic" (e.g. how network instance config is mapped to dnsmasq configlet) and not interaction with external systems (like kernel, dnsmasq, etc.). That I would leave to integration tests that usually run inside containers/VMs as to not interfere and depend on the host and to ensure that everything is properly cleaned when they stop.
For developers this test should be harmless because it will be skipped due to a lack of privileges anyway. However, inside github workers the IP attached to the bridge could cause some interference. But we can leave it and we will see if there will be any issues...

Reply:
I can also make it an "Example" - https://pkg.go.dev/testing#hdr-Examples

Reply:
Personally I would vote for turning `TestRunDnsmasq` into example. For UT it is quite complex and affects host networking, but for development and manual testing it can be actually very useful for the purpose of trying different dnsmasq options, observing how it behaves, etc.

Reply:
I tried to turn it into an "Example", but unfortunately it is not possible to skip examples, so it would then fail in case it cannot create a bridge device.

Reply:
And it will be skipped if the tests are run with `-short` - https://github.com/lf-edge/eve/pull/3028/files#diff-ed489fe6ff04b22d14fbd35a110fc8fbc8fd3aa82ff26ba1646e75cfdb3f90efR381

Reply:
OK, makes sense